staging/src/k8s.io/api/storage/v1beta1/types.go

```diff
@@ -209,3 +209,174 @@ type VolumeError struct {
+	// attachRequired indicates this CSI volume driver requires an attach
+	// operation (because it implements the CSI ControllerPublishVolume()
+	// method), and that Kubernetes should call attach and wait for any attach
+	// operation to complete before proceeding to mounting.
+	// If value is not specified, default is false -- meaning attach will not be
+	// called.
+	// +optional
+	AttachRequired bool `json:"attachRequired,omitempty" protobuf:"varint,1,opt,name=attachRequired"`
+
+	// If specified, podInfoRequiredOnMount indicates this CSI volume driver
+	// requires additional pod information (like podName, podUID, etc.) during
+	// mount operations.
+	// If value is not specified, pod information will not be passed on mount.
+	// If value is set to a valid version, Kubelet will pass pod information as
+	// VolumeAttributes in the CSI NodePublishVolume() calls.
+	// Supported versions:
+	// Version "v1" will pass the following ValueAttributes
+	// "csi.storage.k8s.io/pod.name": pod.Name
+	// "csi.storage.k8s.io/pod.namespace": pod.Namespace
+	// "csi.storage.k8s.io/pod.uid": string(pod.UID)
+	// +optional
+	PodInfoOnMountVersion string `json:"podInfoOnMountVersion,omitempty" protobuf:"bytes,2,opt,name=podInfoOnMountVersion"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSINodeInfo holds information about all CSI drivers installed on a node.
+type CSINodeInfo struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata.name must be the Kubernetes node name.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec is the specification of CSINodeInfo
+	Spec CSINodeInfoSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status is the current status of CSINodeInfo
+	Status CSINodeInfoStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// CSINodeInfoSpec holds information about the specification of all CSI drivers installed on a node
+type CSINodeInfoSpec struct {
+	// drivers is a list of specifications of CSIDriverInfo
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Drivers []CSIDriverInfoSpec `json:"drivers" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=drivers"`
+}
+
+// CSINodeInfoStatus holds information about the status of all CSI drivers installed on a node
+type CSINodeInfoStatus struct {
+	// drivers is a list of the statuses of CSIDriverInfo
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Drivers []CSIDriverInfoStatus `json:"drivers" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=drivers"`
+}
+
+// CSIDriverInfoStatus holds information about the status of one CSI driver installed on a node
+type CSIDriverInfoStatus struct {
+	// name is the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
```

Comment:
The entries in spec and status for the same driver should always have the same name. If they don't match they are not the same driver. We need them in each so that we can tell what driver they are for. Since Spec's and Statuses are just represented as lists we need to be able to tell which driver each Spec and each Status is for.

Reply:
Did we ever entertain the idea of making the list of drivers a map instead? Then we could remove the validation that checks for duplicate driver entries in a list.

Reply:
I don't think it was considered in the original design, it might be a good choice. I am not familiar enough with k8s API design to make a call either way.

Reply:
@xing-yang wdyt of making the drivers list a map instead of a list? Then we can remove a lot of the validation logic around duplicate entries

Reply:
Sadly, we don't do maps in the k8s API.  Places like this would be nice, but no.

Reply:
Can you clarify the no maps policy? I see some maps defined in the core api, like [here](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/apis/core/types.go#L2550)